### PR TITLE
Update BPF lexer

### PR DESCRIPTION
**Pull request by pchaigno**
This pull request updates the BPF lexer to support several constructs from the two main disassemblers, bpftool and objdump:
- `w4` syntax for registers used in 32bit ALU instructions.
- `map[id:671]` to refer to pointer to the map with id 671.
- `call bpf_skb_load_bytes#7684912` to refer to a call to helper bpf_skb_load_bytes.
- line numbers
- `(bf)` and `61 12 04 00 00 00 00 00` for the hexadecimal encoding of the opcode and instruction respectively.

I'm adding basic support for the hexadecimal output (`Generic` group) so that it doesn't break highlighting. By supporting this, we make it easier for users to directly highlight the output of bpftool and objdump. I've updated the sample file with raw outputs from both bpftool and objdump.

### CODE REVIEW

File: lib/rouge/lexers/bpf.rb [hunk 1]
```diff
@@ -13,25 +13,39 @@ class BPF < RegexLexer
...
-        be16 be32 be64 exit lock
+        be16 be32 be64 exit lock map
       ).join('|')
 
       state :root do
+        # Line numbers and hexadecimal output from bpftool/objdump
+        rule %r/(\d+)(:)(\s+)(\(\h{2}\))/i do
+          groups Generic::Lineno, Punctuation, Text::Whitespace, Generic
+        end
+        rule %r/(\d+)(:)(\s+)((\h{2} ){8})/i do
```
Comment (pyrmont) [inline lib/rouge/lexers/bpf.rb]:
```suggestion
        rule %r/(\d+)(:)(\s+)((?:\h{2} ){8})/i do
```

We don't actually want this parenthesis to be a capture group, right?

File: lib/rouge/lexers/bpf.rb [hunk 2]
```diff
@@ -13,25 +13,39 @@ class BPF < RegexLexer
...
 
       MISC_KEYWORDS = %w(
-        be16 be32 be64 exit lock
+        be16 be32 be64 exit lock map
       ).join('|')
 
       state :root do
+        # Line numbers and hexadecimal output from bpftool/objdump
+        rule %r/(\d+)(:)(\s+)(\(\h{2}\))/i do
+          groups Generic::Lineno, Punctuation, Text::Whitespace, Generic
```
Comment (pyrmont) [inline lib/rouge/lexers/bpf.rb]:
```suggestion
          groups Generic::Lineno, Punctuation, Text::Whitespace, Literal::Number::Hex
```

Is there a reason not to use the hexadecimal token?

File: lib/rouge/lexers/bpf.rb [hunk 3]
```diff
@@ -13,25 +13,39 @@ class BPF < RegexLexer
...
+        end
+        rule %r/(\d+)(:)(\s+)((\h{2} ){8})/i do
+          groups Generic::Lineno, Punctuation, Text::Whitespace, Generic
```
Comment (pyrmont) [inline lib/rouge/lexers/bpf.rb]:
```suggestion
          groups Generic::Lineno, Punctuation, Text::Whitespace, Literal::Number::Hex
```

Same issue as above.

File: lib/rouge/lexers/bpf.rb (hunk 1 above)
Comment (pchaigno) [reply lib/rouge/lexers/bpf.rb]: Ah, I forgot to add it. Fixed (with rebase). Thanks!

File: lib/rouge/lexers/bpf.rb (hunk 2 above)
Comment (pchaigno) [reply lib/rouge/lexers/bpf.rb]: I'm not sure they should be highlighted. They're not very relevant to the output. I just want to ensure their presence doesn't break highlighting of the actual bytecode (the rest of the line).

Comment (pyrmont) [reply lib/rouge/lexers/bpf.rb]: Gotcha.

File: lib/rouge/lexers/bpf.rb (hunk 3 above)
Comment (pyrmont) [reply lib/rouge/lexers/bpf.rb]: See above comments.

**Comment by pyrmont** [comment]
@pchaigno Thanks for adding this support and I'm sorry I didn't get it into master before cutting the v3.12.0 release :(

We're about to move to a monthly release schedule for new versions of Rouge so I'm afraid there'll be a bit of a wait. The next version is scheduled to be released on 12 November.

Thanks again for contributing to Rouge!

**Comment by pchaigno** [comment]
> We're about to move to a monthly release schedule for new versions of Rouge so I'm afraid there'll be a bit of a wait. The next version is scheduled to be released on 12 November.

No problem. It takes many months to update the dependencies in [GitHub Pages gem](https://github.com/github/pages-gem) anyway.

**Pull request merged**